### ISSUE TITLE
[Due for payment 2025-03-28] Render shortened mentions with blue/green outline in live markdown preview

### ISSUE THREAD

**Issue by puneetlath** [Expensify/App#38025]
# Problem

Our markdown rendering renders mentions with a blue or green outline within the chat body like this: 
<img width="401" alt="Screenshot 2024-03-10 at 1 10 07 PM" src="https://github.com/Expensify/App/assets/1371996/5dcd9fb6-57a5-435e-9992-921e12338e31">

We also have "live markdown preview" in the composer which is supposed to give a live preview of how the text will be rendered like this: 
<img width="376" alt="Screenshot 2024-03-10 at 1 12 53 PM" src="https://github.com/Expensify/App/assets/1371996/5f7acc7b-a8f0-4fec-b244-c8ba273f65a9">

However, this doesn't currently happen for shortened mentions: 
<img width="379" alt="Screenshot 2024-03-10 at 1 18 30 PM" src="https://github.com/Expensify/App/assets/1371996/50c09d69-ba5b-4242-87a7-87edcf1ac2e2">

# Solution
Add shortened mentions to the live markdown preview rendering that happens in the composer. 

Note for potential contributors: live markdown preview currently only exists on mobile apps. 

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01b3ecb2640f3bf7a3</li>
        <li>Upwork Job ID: 1766890615417413632</li>
        <li>Last Price Increase: 2024-03-10</li>
<li>Automatic offers: </li>
<ul>
<li>shubham1206agra | Contributor | 105292539</li>
</ul></li>
    </ul>
</details>

**Comment by OskarMast** [comment]
No problem. Done

**Comment by SzymczakJ** [comment]
Hey! I’m Jakub Szymczak from Software Mansion, an expert agency, and I’d like to work on this issue!

**Comment by puneetlath** [comment]
@SzymczakJ any thoughts so far?

**Comment by SzymczakJ** [comment]
@puneetlath working on it. We would like to solve this issue without hindering app performance, to do this we have to do some things around react-native-live-markdown(asked @tomekzaw who is react-native-live-markdown creator and it's also his opinion). Without that improvement of react-native-live-markdown the short mentions are going to hinder app performance, but when we complete it, then this issue and many bugs around live-markdown are going to be easily solvable.
 
One downside is that this change in live markdown library is not so easily done, so is it possible to postpone this task and do it after we adjust of live-markdown library?

**Comment by thienlnam** [comment]
Yeah I think that's fine, this doesn't seem that urgent - what adjustments would we need to make in the live-markdown library?

**Comment by puneetlath** [comment]
>is it possible to postpone this task and do it after we adjust of live-markdown library?

I agree, that seems fine. Is there an issue where this adjustment is being tracked and where we could get a summary of what changes are needed?

**Comment by tomekzaw** [comment]
> what adjustments would we need to make in the live-markdown library?

@thienlnam `react-native-live-markdown` needs to parse and format Markdown string on the UI thread while typing. Currently, we're running ExpensiMark from a separate bundle which is embedded in the library. This approach allowed us to quickly deliver the initial version for native but as it turns out it also has a few limitations that we would need to eliminate as a long-term improvement, for instance:

* **Syncing expensify-common version between E/App and Live Markdown.** Currently, Live Markdown uses its own version of ExpensiMark which may be different from the one used in E/App. This may lead to inconsistencies in formatting the messages between the composer and chat history. Also, recently ExpensiMark has been updated quite frequently and the upgrade process of expensify-common is more complicated than expected since we need to bump the version of ExpensiMark separately in E/App and Live Markdown library, and then also bump Live Markdown version in E/App.

* **Customizing Markdown parser with app state**. Currently, the parsing and formatting algorithm is hard-coded into the library so it always behaves the same and does not depend on app state. However, in some cases we would like the current app state to affect the results of parsing or formatting, e.g. short mentions (where parsing depends on the list of already contacted/mentioned users) or user mentions (where formatting depends on whether the current user is mentioned or somebody else).

The solution in my mind is to get rid of the separate bundle and move existing JS parser/formatter algorithm implementation into a [worklet](https://docs.swmansion.com/react-native-reanimated/docs/fundamentals/glossary/#worklet). This way, its source code can be embedded in the same React bundle (so it will always use the same version of ExpensiMark as the E/App does). As for parametrizing parsing/formatting logic with app state, we can achieve it using shared state, i.e. [shared values](https://docs.swmansion.com/react-native-reanimated/docs/core/useSharedValue/).

**Comment by tomekzaw** [comment]
@puneetlath We're not actively working on this particular feature but we are doing some improvements in Live Markdown that will allow us to implement these features (currently it's not really possible).

**Comment by puneetlath** [comment]
Ah cool, got it. Where can I follow along with that @tomekzaw?

**Comment by tomekzaw** [comment]
I've created an issue with a detailed description in the Live Markdown repo: 
* https://github.com/Expensify/react-native-live-markdown/issues/317

**Comment by puneetlath** [comment]
Great, thanks! I'll follow along there.

**Comment by puneetlath** [comment]
Still on hold.

**Comment by quinthar** [comment]
Why is this on hold?  When will it come off hold?  This feels very broken

**Comment by puneetlath** [comment]
@tomekzaw were you able to find someone to take on the linked issue above?

**Comment by tomekzaw** [comment]
@puneetlath We've discussed the plan and one of the prerequisites would be to eliminate Underscore.js and jQuery from ExpensiMark as they cannot be workletized. How can we approach this (internally or externally)?

**Comment by puneetlath** [comment]
That's interesting. We can have someone do it, since it's an open source repo. Can you post a Problem/Solution in #expensify-open-source and then we'll hire someone to do it assuming we agree. Thanks!

**Comment by thienlnam** [comment]
Any progress on this one https://github.com/Expensify/react-native-live-markdown/issues/317 cc @tomekzaw?

**Comment by tomekzaw** [comment]
Discussing it here: https://swmansion.slack.com/archives/C01GTK53T8Q/p1716235671381059

**Comment by puneetlath** [comment]
Still on hold, but we're making progress.

**Comment by thienlnam** [comment]
Discussion here: https://expensify.slack.com/archives/C01GTK53T8Q/p1716235671381059

Seems like we're currently holding on https://github.com/Expensify/App/issues/42247

**Comment by puneetlath** [comment]
Still on hold, but the issues we're held on are making progress.

**Comment by tomekzaw** [comment]
Waiting for https://github.com/Expensify/react-native-live-markdown/issues/317

Latest update here https://github.com/Expensify/react-native-live-markdown/issues/317#issuecomment-2196474651

**Comment by thienlnam** [comment]
Still held

**Comment by puneetlath** [comment]
Still on hold for https://github.com/Expensify/react-native-live-markdown/issues/317

**Comment by thienlnam** [comment]
Same ^

**Comment by thienlnam** [comment]
Looks like the above is closed to getting merged, should be off hold soon

**Comment by puneetlath** [comment]
@tomekzaw can this come off hold now and be worked on?

**Comment by tomekzaw** [comment]
@puneetlath Yes, we can remove hold on this one.

**Comment by puneetlath** [comment]
Great and will @SzymczakJ still be the person to work on it?

**Comment by tomekzaw** [comment]
Someone from SWM will handle this task as this requires some extra thoughts on how to pass app state to Markdown parser efficiently. Let me check internally who's the lucky person and I will let you know.

**Comment by tomekzaw** [comment]
Just for the context, we're still waiting for the live-markdown bump to be merged (https://github.com/Expensify/App/pull/53627) so let's hold on that PR for now

**Comment by tomekzaw** [comment]
Someone from SWM will work on this feature

**Comment by Kicu** [comment]
Hey guys I work form SWM, please assign me to this issue and I will start working on it.

**Comment by allroundexperts** [comment]
> Can I take this over as a reviewer here?
> Since I am C+ assigned in https://github.com/Expensify/App/issues/35725.

@shubham1206agra Do you have greater context here? I see that #35725 is just on hold for this issue. I was tagging along this one since March 😢

**Comment by allroundexperts** [comment]
> @allroundexperts Yes since I was following the upstream PR and https://github.com/Expensify/App/issues/35725 is there since February.

Gotcha.

**Comment by Kicu** [comment]
@puneetlath I'm not sure if you're the best person to ask, but I have some questions.

I'm currently looking at the regexes in ExpensiMark used to parse `userMention` (https://github.com/software-mansion-labs/expensify-common/blob/main/lib/ExpensiMark.ts#L365) and I'm trying to add another regex for `shortMention`.
We need ExpensiMark to return these shortened version so that later (in a worklet) we can compare them with your onyx data.

What specific versions of short mentions should we support? I'm assuming the most basic one is: `@john.doe`.
But should I also support:
 - `@justOneWord`?
 - any special characters?
 - are phone numbers affect short mentions in any way?

I guess I use the first part of `CONST.REG_EXP.EMAIL_PART` for this. Should

**Comment by puneetlath** [comment]
We need to support anything that could be before the `@` in an email address. So whatever formatting and special characters are supported for that. I don't think phone numbers matter for short mentions.

**Comment by Kicu** [comment]
Before Christmas update.

For the past few days I've been working on ExpensiMark, to try to understand the current email-parsing regexes and to add `short-mentions`.
Here is a draft PR that shows how this could look like, most of the `ExpensiMark.replace` tests are passing, there are several edge-cases that need further digging.

Please take a look at the PR and more specifically at which tests are failing, because some of these are tricky and need some agreeing what we would consider a short mention.

The current plan would be:
 - allow ExpensiMark to understand and parse short mentions, so that the tag `<mention-short` can be returned 
 - use the output of ☝️in `parseExpensiMark` fn in live-markdown, to be able to create markdown with possible mentions
 - in a worklet, inside `E/App` pass list of possible mentions (based on users in Onyx) to a parsing function, and verify if any of "possible" mentions is actually a mention, then decorate it

I'm going on Holiday till January 6th, however I should be around to answer any questions around 30/31 Dec.

**Comment by Kicu** [comment]
## In depth implementation details
In order to implement this task we need to modify 3 (sic!) repos: A/App, `expensify-common` (because Expensimark is there) and `react-native-live-markdown`.

### List of PRs (still work in progress, I will update this comment in future):
 - `live-markdown`: https://github.com/Expensify/react-native-live-markdown/pull/592 -> adds the new type of `mention-short` so that `parseExpensiMark` function understands it
 - `expensify-common`: https://github.com/Expensify/expensify-common/pull/824 -> makes expensimark parser understand and return short mentions
 - `E/App`: https://github.com/Expensify/App/pull/54037 -> adds actual handling of short mentions inside worklet, verifying if the mention is pointing to an actual users login

**Comment by thienlnam** [comment]
expensify-common PR in review - @shubham1206agra could you also take a look at this PR since you're the C+ here? https://github.com/Expensify/expensify-common/pull/824 

Tried to assign you but can't until you comment on it

**Comment by puneetlath** [comment]
Just to make sure I understand how this would work: 

1. If a user types `@` and then chooses `@puneet` from the suggestions, this would then add markdown `<mention-short>puneet</mention-short>` is that right? 
2. If a user types `hi @puneet, how are you?` on their own without using the suggester, this would add markdown of `hi <mention-short>puneet</mention-short>, how are you?` is that right? 
3. The API would then need to handle `<mention-short>` and try to turn those into real mentions. 

Can you help me understand why we are able to render the short mention in the chat body without having `mention-short` but we aren't able to do so in live markdown preview? 

I'm not pushing back against anything. Just trying to understand the what and why of the approach. Thanks!

**Comment by Kicu** [comment]
Thanks for looking into this.

Re 1. Yes correct
Re 2. Yes correct
Re 3. If by `API` you mean the component interface of `react-native-live-markdown` and our parser (the code that calls `ExpensiMark`) then yes. However no backend api (or http calls) are involved in this solution, as this is a frontend-only solution, based on data already existing in onyx

So here's the situation to be the best of my knowledge (but I also got involved only a few weeks ago, I'm missing some previous context, which frankly I thought you might have 😅).

AFAICS there are 2 separate mechanisms responsible for displaying styled user mentions. In the chat body it seems that backend api simply sends us already prepared HTML messages, with the tags like `<mention-user>` already in there. Then we just display them using a library called `'react-native-render-html'`. (see screenshot 1.). 
I have no idea what happens on backend/api and how this whole flow works, as I have no access there. I hoped someone from Expensify team could shed some light.

However we cannot use a similar flow inside composer. In there we use `react-native-live-markdown` which works on markdown representation, not html representation. Until now there was no way for us to verify with actual user data whether text looking like: `@john.doe` is a _legit_ user mention or not. We don't do any api calls, we just parse whatever user typed in.

So now thanks to introducing worklets we can actually verify a user mention (via comparing the mention with users onyx data) and provide a solution.

It is not an ideal situation that 2 mechanisms exist for mentions, but that is the situation we are in.

**Comment by puneetlath** [comment]
Ok got it! I think that makes sense then. I would love to test a POC if possible.

**Comment by Kicu** [comment]
The problem with POC is that we would need to somehow orchestrate together a version of my PR on Expensify/App together with specific changes from `expensify-common`. I'm not sure how I could provide this to you in a reasonable way?

When I was developing this I basically manually put changes from `expensify-common` to my node_modules. 
What can I do to help you test this?

**Comment by puneetlath** [comment]
@shubham1206agra have you been able to test this locally?

**Comment by Kicu** [comment]
I will update this issue later today

**Comment by Kicu** [comment]
I have rebased my PR (https://github.com/Expensify/App/pull/54037) and made sure it works.

In order to simplify testing I have created and commited a temporary patch to `expensify-common`. 
The purpose of this patch is to introduce changes to ExpensiMark that I described in this PR: https://github.com/Expensify/expensify-common/pull/824. Thanks to it you guys don't need to manually modify your local `node_modules` - when the patch applies everything should just work.

We will remove the patch later before merging, as it won't be needed once expensiMark gets merged.
I tested on web and on ios and typing a short mention gets highlighted - did not have time to test on android but I can do so tomorrow.
Please leave any comments and questions you might have :)
CC @shubham1206agra @puneetlath 

<img width="370" alt="Image" src="https://github.com/user-attachments/assets/ca84c719-6a54-4a91-be73-18aa0aa19427" />

**Comment by Kicu** [comment]
Thank you @shubham1206agra I will try to come back to you tomorrow. Since this task is dragging over a long time, I got involved in some other Reports related issues.

**Comment by puneetlath** [comment]
Thanks for the testing steps. I tested it out and it seems to be working well from what I can see.

**Comment by joekaufmanexpensify** [comment]
[Checking in](https://expensify.slack.com/archives/C01SKUP7QR0/p1743172050494139) about payment expectations here as I see there were a few regressions. Will pay early next week!

**Comment by joekaufmanexpensify** [comment]
@shubham1206agra We discussed in Slack and landed on $250 being fair here. There were multiple PRs ([here](https://github.com/Expensify/expensify-common/pull/824) and [here](https://github.com/Expensify/App/pull/54037)), but also a few regressions ([here](https://github.com/Expensify/App/issues/58887) and [here](https://github.com/Expensify/App/issues/58893)), so $250 seems reasonable. Let us know if any concerns on that!

**Comment by joekaufmanexpensify** [comment]
Also, given you were hired here in 2024, we need to pay you in Upwork still for this.

**Comment by joekaufmanexpensify** [comment]
Offer was not accepted on old job, and it has since expired. So I opened a new one [here](https://www.upwork.com/jobs/~021906826918088035074).

**Comment by joekaufmanexpensify** [comment]
@shubham1206agra offer sent for $250!

**Comment by joekaufmanexpensify** [comment]
@shubham1206agra if you could also [propose a regression test](https://github.com/Expensify/App/issues/38025#issuecomment-2744581450), that would be great too. Thank you!

**Comment by joekaufmanexpensify** [comment]
[Discussion regression test](https://expensify.slack.com/archives/C02NK2DQWUX/p1743543924882119)

**Comment by joekaufmanexpensify** [comment]
All set to issue payment! We need to pay $250 to @shubham1206agra for C+ review via Upwork.

**Comment by joekaufmanexpensify** [comment]
@shubham1206agra $250 sent and contract ended!

**Comment by joekaufmanexpensify** [comment]
Upwork job closed.

**Comment by joekaufmanexpensify** [comment]
All set, thanks everyone!
